core/jsAndWasmShared/src/LocalDate.kt

```diff
@@ -68,23 +68,27 @@ public actual fun LocalDate.plus(value: Long, unit: DateTimeUnit.DateBased): Loc
 
 private fun LocalDate.plusNumber(value: Number, unit: DateTimeUnit.DateBased): LocalDate =
         try {
-            when (unit) {
-                is DateTimeUnit.DayBased -> this.value.plusDays(value.toDouble() * unit.days)
-                is DateTimeUnit.MonthBased -> this.value.plusMonths(value.toDouble() * unit.months)
-            }.let(::LocalDate)
+            jsTry {
+                when (unit) {
+                    is DateTimeUnit.DayBased -> this.value.plusDays(value.toDouble() * unit.days)
```

Comment:
The `jsTry` change is extremely intrusive and pollutes all JS code. If I understand correctly what's going on, this is a measure to support the exceptions thrown from pure JS (not Kotlin/JS) code, which seems incompatible with how Kotlin/WASM treats exceptions. Is this correct?

If so, we'd need to remember to surround all the JS code we call with such checks. It is easy to miss, and I don't even know what will happen if we do miss this. How badly will something break? This question needs to be answered in form of a comment somewhere in the code, no matter what solution we end up with.

Here's an idea of how this could be done in a more robust manner:

```kotlin
// jsAndWasmShared
internal expect class JodaTimeLocalDate {
  fun plusDays(days: Number): Number
  // ...
}

class LocalDateTime(value: JodaTimeLocalDate) // instead of `value: jtLocalDate`

// js
internal actual typealias JodaTimeLocalDate = jtLocalDate

// wasm
internal actual value class JodaTimeLocalDate(val value: jtLocalDate) {
  fun plusDays(days: Number): Number = JodaTimeLocalDate(jsTry { value.plusDays(days) })
  // ...
}
```

This way, the shared source set will stay almost the same, and all WASM-specific logic will be in one place where it's easy to review and spot if something's wrong.

What do you think? Is this approach viable?

Reply:
Yeah, when you commonize a code between two platforms with different interop it always looks tricky. This is a price for commonization. 

Yes. In kotlin.wasm platform we have no ability (so far) to catch exceptions thowed from javascript code. This is why I put special wrapper to catch these exceptions inside js world and rethrow it as wasm exceptions in wasm world. For js platform its just a nop.
When I was writing this I had a simple mental rule - every time you want to try-catch exception from js library you should wrap `jsTry` this block.

Your suggest is indeed more readable from a perspective of `jsTry` but one more indirection level makes a mess and boilerplate.

There is several cons for this:
1) Now we have two `Instant` types - `DateTime.Instant` and `jsJoda.Instant` external. With this approach we will have three similar types - `DateTime.Instant`, `jsJoda.Instant` and `WasmWrapper.Instant`.
2) Everywhere we use joda externals we need now make 2 inbox operations, like `jtDuration.between(other.value.jodaInstant, this.value.jodaInstant)`.
3) Everywhere we pass `jsJoda` declarations to `DateTime` declarations we need to pack twice.
4) You need to overlap almost all Joda external hierarchy.
5) It does not solves major problem, because you still need to know which methods should be wrapped with jsTry and which is not. 

Now you can formulate a common rule - any place I put `try-catch` for js exception whould be wrapped with `jsTry` also. Not sure that wrappers can do better.

But:
If you still think that it can do better, I can try to do that but for now it does't seem any better for me.

Reply:
I'm ok with the current jsTry approach, we could refactor it later if we find it disadvantageous.

Reply:
Plus it is hard to actualize class with value class. Moreover actual class cannot be open so you cannot build hierarchy with value classes. I believe that code become more and more unreadable if we wish to escape using jsTry with suggested approach. As far I do this refactoring as more hard it to read and support.

Reply:
Here I made half of work to separate Joda API from external declarations. This code will be doubled to provide support for the Wasm.
https://github.com/Kotlin/kotlinx-datetime/commit/e78597df98eb0a0aade70f43e52c40299ce5ddd8

Reply:
> You cant make typealias actual in js for these external declarations.

If this is true, I agree that this is a deal-breaker. The ridiculous boilerplate would be a greater evil than `jsTry`.

But I don't understand why this is true. I just tried it in your `wasm_add_layer` branch, and it seems to work:

```kotlin
// core/jsAndWasmShared/src/JsJodaPlatform.kt
public expect class JodaTimeZoneId {
    public fun id(): String
    public fun normalized(): JodaTimeZoneId

    public companion object {
        public fun systemDefault(): JodaTimeZoneId
        public fun of(zoneId: String): JodaTimeZoneId
    }
}

// core/js/src/JsJodaPlatform.kt
actual typealias JodaTimeZoneId = kotlinx.datetime.internal.JSJoda.ZoneId
```
At least in the IDE, the only red place in this code is `Expected class 'JodaTimeZoneId' has no actual declaration in module Kotlin-DateTime-library.kotlinx-datetime.wasmJsMain for Wasm`. Are there any fundamental reasons why this approach wouldn't work everywhere? Note that we don't need Joda Time's class hierarchies, only the final classes.

Reply:
The only way I found to do what you suggest is here:
https://github.com/Kotlin/kotlinx-datetime/blob/b3dc7699fa116292890be6971e7904cc50fbdc63/core/wasmJs/src/JsJodaPlatform.kt#L15

I cannot use `value` classes because you cannot actualize value class with non-value classes and vice versa.

Reply:
I double-checked, and it looks like it's true, you can't do `actual value class` for an `expect class`. Interesting.

Yes, I think it's the better separation of responsibilities, thank you for researching this! @ilya-g, do you have anything against this approach?

Reply:
If @ilya-g agree that it is better then I will complete this refactoring next week. Thanks!

Reply:
This refactoring would complicate code and prolong the review. I would rather go with the current `jsTry` in shared code approach and postpone this refactoring.

Reply:
> would complicate code and prolong the review

Funnily enough, I disagree with both points. Still, we have "one subjective opinion + effort to refactor" on one side and "one subjective opinion" on the other, so since we are split on this, not doing the refactoring wins in any case. Let's drop this line of inquiry.

Reply:
Some miscommunication probably happened: I agreed not to have this refactoring. Sorry if I worded this incorrectly!

Reply:
Ohh, I missed this post. :(
So, need I revert this?

Reply:
Yes, please do, if you want this review to be completed sooner than later

Reply:
@ilya-g, why do you believe that this approach would take longer to review? With this separation, we barely have to look at the shared code; the changes there are trivial. We only have to look through a small list of definitions and check that they are wrapped.

What *does* make this PR difficult to review now is the amount of code that's touched but not meaningfully changed. This needs to be fixed. After that, the changes would be self-contained.

Reply:
I reduced unnecessary touched code. If you still want me to remove wrappers then I will.